src/lib/device/journal_vdev.cpp

```diff
@@ -559,54 +574,108 @@ void JournalVirtualDev::Descriptor::truncate(off_t truncate_offset) {
+    // cover_offset is a logical offset.
+    index = 0;
+    off_t tail_off =
+        static_cast< off_t >(data_start_offset() + m_write_sz_in_total.load(std::memory_order_relaxed)) + m_reserved_sz;
+    auto chunk_size = m_vdev.info().chunk_size;
+    LOGINFOMOD(journalvdev, "Truncate begin truncate {} desc {}", to_hex(truncate_offset), to_string());
+
+#ifdef _PRERELEASE
+    for (auto it = m_journal_chunks.begin(); it != m_journal_chunks.end(); ++it) {
+        auto chunk = *it;
+        auto* private_data = r_cast< JournalChunkPrivate* >(const_cast< uint8_t* >(chunk->user_private()));
+        LOGINFOMOD(journalvdev, "log_dev={} chunk_id={} is_head={}", m_logdev_id, chunk->chunk_id(),
+                   private_data->is_head)
+    }
+#endif
+
+    off_t cover_offset = sisl::round_down(data_start_offset(), chunk_size);
+    auto total_num_chunks = m_journal_chunks.size();
     for (auto it = m_journal_chunks.begin(); it != m_journal_chunks.end();) {
         auto chunk = *it;
-        start += chunk->size();
+        off_t end_of_chunk = m_vdev.get_end_of_chunk(chunk);
+        off_t chunk_hole = 0;
+        if (index == total_num_chunks - 1) {
+            // If its the last chunk, only read upto the tail_offset
+            // There are no holes in the last chunk.
+            cover_offset += (tail_off % chunk_size);
+        } else {
+            // For other chunks take the whole size.
+            cover_offset += chunk_size;
+            chunk_hole = (chunk_size - end_of_chunk);
+        }
 
-        // Also if its the last chunk and there is no data after truncate, we release chunk.
-        auto write_sz_in_total = m_write_sz_in_total.load(std::memory_order_relaxed);
-        if (start >= truncate_offset) { break; }
+        index++;
 
-        m_total_size -= chunk->size();
-        it = m_journal_chunks.erase(it);
+        // Check if the offset is inside the truncate offset. We also check if the truncate offset lies
+        // between end_of_chunk - chunk_size. If either condition satifies, we release the chunks.
+        if (cover_offset <= truncate_offset || ((cover_offset - chunk_hole) <= truncate_offset)) {
+            m_total_size -= chunk->size();
+            it = m_journal_chunks.erase(it);
 
-        // Clear the private chunk data before adding to pool.
-        auto* data = r_cast< JournalChunkPrivate* >(const_cast< uint8_t* >(chunk->user_private()));
-        *data = JournalChunkPrivate{};
-        m_vdev.update_chunk_private(chunk, data);
+            // Release the chunk back to pool.
+            LOGINFOMOD(journalvdev,
+                       "Released chunk_id={} log_dev={} cover={} truncate_offset={} tail={} end_of_chunk={} desc {}",
+                       chunk->chunk_id(), m_logdev_id, to_hex(cover_offset), to_hex(truncate_offset), to_hex(tail_off),
+                       m_vdev.get_end_of_chunk(chunk), to_string());
+            m_vdev.release_chunk_to_pool(chunk);
```

Comment:
have we thought through if before and after this call (with persistence to disk), everything will be booted up okay?
And we need to have a flip here to simulate this crash and verify boot recovery is fine.

Reply:
Added this in test case in the logstore crash recovery doc.

Reply:
which doc are you referring to? I think better way is to create an issue, otherwise we will forget which of the tests are there and which are TODOs.

Reply:
We have a issue for this https://github.com/eBay/HomeStore/issues/379

Reply:
https://docs.google.com/document/d/1ooA9w_kkg6P9G9l72xDuZpLJ2sRp1mbhCMFwkmRkfRg/edit?pli=1